### PR TITLE
SC-687-crosscontract-invoke

### PR DESCRIPTION
**Pull request by potan**

### CODE REVIEW

File: node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala [hunk 1]
```diff
@@ -2172,9 +2181,1221 @@ class InvokeScriptTransactionDiffTest
...
+
+      compileContractFromExpr(expr, V5)
+    }
+    val scenario =
+      for {
+        master  <- accountGen
+        invoker <- accountGen
+        service <- accountGen
+        ts      <- timestampGen
+        fee     <- ciFee(1)
```
Comment (xrtm000) [inline node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala]: Why is it enough despite issuing via action?
See line 2330:
https://github.com/wavesplatform/Waves/pull/3316/files#diff-7a87dd14a82b2cd0e58fddefc26230665e583d933ab5ba4da3ea943f13f4a7f1R2330

File: node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptTransactionDiff.scala [hunk 2]
```diff
@@ -100,23 +102,19 @@ object InvokeScriptTransactionDiff {
...
                   tthis,
                   directives,
-                  tx.id(),
-                  disableThisWhenCalledByAlias
+                  tx,
+                  dAppAddress,
+                  pk,
+                  dAppAddress,
+                  runsLimit,
+                  13
```
Comment (xrtm000) [inline node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptTransactionDiff.scala]: What is it? Should be it tested?

File: node-it/src/test/scala/com/wavesplatform/it/api/model.scala [hunk 3]
```diff
@@ -416,14 +416,32 @@ object AssetPairResponse {
...
+  implicit val callFormat: Reads[Call] = Json.reads[Call]
+}
+
+case class AttachedPayment(asset: Option[String], amount: Long)
+object AttachedPayment {
+  implicit val attachedPaymentFormat: Reads[AttachedPayment] = Json.reads[AttachedPayment]
+}
+
+case class Invokation(dApp: String, call: Call, payments: Seq[AttachedPayment], stateChanges: StateChangesDetails)
+object Invokation {
```
Comment (xrtm000) [inline node-it/src/test/scala/com/wavesplatform/it/api/model.scala]:
```suggestion
object Invocation {
```

File: node/src/main/scala/com/wavesplatform/state/InvokeScriptResult.scala [hunk 4]
```diff
@@ -56,14 +67,23 @@ object InvokeScriptResult {
...
   implicit val jsonFormat         = Json.writes[InvokeScriptResult]
 
   implicit val monoid = new Monoid[InvokeScriptResult] {
     override val empty: InvokeScriptResult =
       InvokeScriptResult.this.empty
 
-    override def combine(x: InvokeScriptResult, y: InvokeScriptResult): InvokeScriptResult =
-      InvokeScriptResult(x.data ++ y.data, x.transfers ++ y.transfers)
+    override def combine(x: InvokeScriptResult, y: InvokeScriptResult): InvokeScriptResult = {
+      InvokeScriptResult(/*x.data ++ y.data, x.transfers ++ y.transfers,*/ invokes = x.invokes ++ y.invokes)
```
Comment (xrtm000) [inline node/src/main/scala/com/wavesplatform/state/InvokeScriptResult.scala]: Should it be saved unchanged?

**Review by xrtm000 — CHANGES_REQUESTED** [review]

File: node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala (hunk 1 above)
Comment (potan) [reply node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala]: I use NFT

Comment (xrtm000) [reply node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala]: If set Issue action as non-NFT there still will be no an error

File: node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptTransactionDiff.scala (hunk 2 above)
Comment (potan) [reply node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptTransactionDiff.scala]: It limits nested Invoke. Tests "Infilite recurcive crosscontract call".

File: node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala (hunk 1 above)
Comment (potan) [reply node/src/test/scala/com/wavesplatform/state/diffs/ci/InvokeScriptTransactionDiffTest.scala]: Fixed

**Pull request merged**